cg_lims/EPPs/files/femtopulse_csv.py

```diff
@@ -1,87 +1,87 @@
-                    "error": "This position is not possible for the run.",
-                }
-            )
-
-    # Prints out error message(s)
-    if failed_samples:
-        all_errors = ""
-        for sample in failed_samples:
-            error_message: str = (
-                f"Sample {sample['artifact_name']} in position {sample['parsed_well']}: {sample['error']}"
-            )
-            all_errors = all_errors + " " + error_message
-        raise InvalidValueError(f"Errors found: {all_errors}")
-
-    # The ladder will always be in well A12
-    DATAFRAME["sample names"].iloc[-1] = "ladder"
-
-    # Create the csv file
-    DATAFRAME.index = range(1, len(DATAFRAME) + 1)
-    DATAFRAME.to_csv(Path(file), index=True, header=False)
+    index, sample position and sample name, blank or ladder."""
+
+    # Get well positions and sample names and sort by well position
+    samples_and_positions: tuple = [
+        (get_artifact_well(artifact), get_sample_artifact_name(artifact)) for artifact in artifacts
+    ]
+    samples_and_positions.sort(key=lambda x: (x[0][0], int(x[0][1:])))
+
+    # Check that no sample is in position 12, where the ladders should be.
+    invalid_samples: List = [
+        sample for position, sample in samples_and_positions if position.endswith("12")
+    ]
+    if invalid_samples:
+        raise InvalidValueError(
+            f"Sample(s) {invalid_samples} in position 12, which is reserved for the ladders."
+        )
+
+    # Get all unique rows needed based on sample positions and number of ladders
+    all_rows: List = ["A", "B", "C", "D", "E", "F", "G", "H"]
+    rows_needed: List = sorted(set(position[0] for position, _ in samples_and_positions))
+    for row in all_rows:
+        if len(rows_needed) >= num_of_ladders:
+            break
+        if row not in rows_needed:
+            rows_needed.append(row)
+    rows_needed.sort()
+
+    # Create a list of positions, sample names, blanks and ladders based on the rows needed for the run.
+    sample_position_layout: List = []
+    for row in rows_needed:
+        row_samples: List = [
+            (position, sample)
+            for position, sample in samples_and_positions
+            if position.startswith(row)
+        ]
+        # Add existing samples for positions 1-11
+        for position in range(1, 12):
+            well = f"{row}{position}"
+            sample = next((samp for pos, samp in row_samples if pos == well), "")
+            sample_position_layout.append((well, sample))
+        # Add ladder or empty position for position 12
+        ladder: str = "ladder" if rows_needed.index(row) < num_of_ladders else ""
+        sample_position_layout.append((f"{row}12", ladder))
+
+    # Create and write dataframe
+    df = pd.DataFrame(sample_position_layout, columns=["well positions", "sample names"])
+    df.index = range(1, len(df) + 1)
+    df.to_csv(Path(file), index=True, header=False, sep=";")
```

Comment:
This is probably where I would place the error message regarding ladders

```suggestion
    df.to_csv(Path(file), index=True, header=False, sep=";")
    
    if len(rows_needed) > num_of_ladders:
        raise InvalidValueError(
            f"Warning: The number of populated sample rows (len(rows_needed)) "
            f"exceed the amount of ladders chosen (num_of_ladders)."
        )
```